cpp/src/arrow/array/builder_dict.h

```diff
@@ -724,6 +747,7 @@ using BinaryDictionaryBuilder = DictionaryBuilder<BinaryType>;
 using StringDictionaryBuilder = DictionaryBuilder<StringType>;
 using BinaryDictionary32Builder = Dictionary32Builder<BinaryType>;
 using StringDictionary32Builder = Dictionary32Builder<StringType>;
+using BinaryDictionary64Builder = Dictionary64Builder<LargeBinaryType>;
```

Comment:
@arthurpassos This seems to conflate two different concerns:
1) being allowed to fit more than 2^31 data bytes in a dictionary (for example more than 2^21 strings of 1kB each)
2) being allowed to fit more than 2^31 strings in a dictionary

While 1) already seems quite unlikely to me, 2) seems _extremely_ unlikely (shouldn't you just use plain encoding instead?). Do you really have a use case for it?

Reply:
Hi @pitrou. First of all, thanks for looking into this.

I am trying to fix the issue described in https://github.com/apache/arrow/issues/32723. It's an issue that pops up when data of complex data structures end up being chunked. The goal of this PR is to introduce a setting that'll allow the use of LARGE* variants of string / binary types to avoid chunking as suggested by @emkornfield.

I simply followed my intuiton that if the non-large types rely on 32bits, the large type would rely on 64bits.

Reply:
Ok, but why would you use a dictionary builder in these situations?

Reply:
Tbh, I do not have the exact answer for you. I am very new to this code base.

My best bet would be: I had to create a LARGE* version of `ByteArrayChunkedRecordReader`. The `ByteArrayChunkedRecordReader` class contains a `::arrow::BinaryDictionary32Builder` instance, so I thought the `Large` version would require a 64 bit version of it.

Do you think it's not actually necessary?

Reply:
As in the LARGE version of ByteArrayChunkedReader would still be functional and be able to read a file generated with the below python script?

```
import pyarrow as pa
import pyarrow.parquet as pq

arr = pa.array([[("a" * 2**30, 1)]], type = pa.map_(pa.string(), pa.int32()))
arr = pa.chunked_array([arr, arr])
tab = pa.table({ "arr": arr })

pq.write_table(tab, "test.parquet")
```

Reply:
@emkornfield What would you suggest here? It seems to me that if a large binary type is desired, it's probably not a good idea to use dictionary encoding...

Reply:
@pitrou I have just removed it, it seems like it works without it.

Regarding your list of scenarios for a 64bit dictionary, wouldn't the need to store a single string with more than 2^31 bits also be an option?

Reply:
Parquet does not allowing storing individual strings larger than 31 bits:
https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L424

Reply:
I see. Ok, please let me know what are your conclusions on the 64bit matter.

Reply:
Sorry, catching up on reviews, is there still an open question here?

Reply:
I don't think so. Initially I had implemented the `LargeBinaryDictionaryBuilder` to be 64bit based, but it seems like the 32bit is enough. But please share your thoughts on this